File: todolist/src/types/lifeEvent.ts
/** 预设分类定义 */
export interface CategoryPreset {
  key: string;
  label: string;
  color: string;
}

/** Rust 返回的原始事件数据 */
export interface LifeEvent {
  id: string;
  title: string;
  description: string;
  start_date: string;
  end_date: string | null;
  category: string;
  color?: string;
  start_precision: string;   // 'month' | 'day'
  end_precision: string;     // 'month' | 'day'
  is_highlighted: boolean;
  created_at: string;
  updated_at: string;
}

/** 创建参数 */
export interface CreateLifeEventParams {
  title: string;
  description?: string;
  start_date: string;
  end_date?: string | null;
  category?: string;
  color?: string;
  start_precision?: string;
  end_precision?: string;
  is_highlighted?: boolean;
}

/** 更新参数 */
export interface UpdateLifeEventParams {
  title?: string;
  description?: string;
  start_date?: string;
  end_date?: string | null;
  category?: string;
  color?: string;
  start_precision?: string;
  end_precision?: string;
  is_highlighted?: boolean;
}

/** 关联链接 */
export interface LifeEventLink {
  id: string;
  life_event_id: string;
  entity_type: string;   // 'task' | 'paper' | 'experiment' | 'countdown' | 'city' | 'asset'
  entity_id: string;
  label: string;
  created_at: string;
}

/** 创建链接参数 */
export interface CreateLifeEventLinkParams {
  life_event_id: string;
  entity_type: string;
  entity_id: string;
  label?: string;
}

/** 自动统计 */
export interface LifeEventStats {
  task_count: number;
  paper_count: number;
  experiment_count: number;
}

/** 前端增强展示类型 */
export interface LifeEventDisplay extends LifeEvent {
  durationText: string;     // "1年9个月" | "至今（已过 2年3个月）"
  isOngoing: boolean;       // end_date === null
  categoryColor: string;    // 最终颜色（分类默认色或自定义色）
  categoryLabel: string;    // 分类中文名
  startPrecision: string;
  endPrecision: string;
}

File: todolist/src-tauri/src/models/life_event.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LifeEvent {
    pub id: String,
    pub title: String,
    pub description: String,
    pub start_date: String,
    pub end_date: Option<String>,
    pub category: String,
    pub color: Option<String>,
    pub start_precision: String,
    pub end_precision: String,
    pub is_highlighted: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateLifeEventParams {
    pub title: String,
    #[serde(default)]
    pub description: Option<String>,
    pub start_date: String,
    pub end_date: Option<String>,
    #[serde(default = "default_category")]
    pub category: String,
    pub color: Option<String>,
    #[serde(default = "default_precision")]
    pub start_precision: String,
    #[serde(default = "default_precision")]
    pub end_precision: String,
    #[serde(default)]
    pub is_highlighted: bool,
}

fn default_precision() -> String { "month".into() }
fn default_category() -> String { "other".into() }

#[derive(Debug, Deserialize)]
pub struct UpdateLifeEventParams {
    pub title: Option<String>,
    pub description: Option<String>,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
    pub category: Option<String>,
    pub color: Option<String>,
    pub start_precision: Option<String>,
    pub end_precision: Option<String>,
    pub is_highlighted: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LifeEventLink {
    pub id: String,
    pub life_event_id: String,
    pub entity_type: String,
    pub entity_id: String,
    pub label: String,
    pub created_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateLifeEventLinkParams {
    pub life_event_id: String,
    pub entity_type: String,
    pub entity_id: String,
    #[serde(default)]
    pub label: Option<String>,
}

/// Auto-statistics result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LifeEventStats {
    pub task_count: i64,
    pub paper_count: i64,
    pub experiment_count: i64,
}
